Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+/*
+Low level interface to Meta's zstd library for use in the `zstd` Python library.
+
+Original implementation by Ma Lin with patches by Rogdham.
+
+Refactored for the CPython standard library by Emma Harper Smith.
+*/
+
+#ifndef Py_BUILD_CORE_BUILTIN
+#  define Py_BUILD_CORE_MODULE 1
+#endif
+
+#include "_zstdmodule.h"
+
+/*[clinic input]
+module _zstd
+
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=4b5f5587aac15c14]*/
+#include "clinic/_zstdmodule.c.h"
+
+/* --------------------------
+     Module level error handling
+   -------------------------- */
+
+/* Format error message and set ZstdError. */
+void
+set_zstd_error(const _zstd_state* const state,
+               const error_type type, const size_t zstd_ret)
+{
+    char buf[128];
+    char *msg;
+    assert(ZSTD_isError(zstd_ret));
+
+    switch (type)
+    {
+    case ERR_DECOMPRESS:
+        msg = "Unable to decompress zstd data: %s";
+        break;
+    case ERR_COMPRESS:
+        msg = "Unable to compress zstd data: %s";
+        break;
+    case ERR_SET_PLEDGED_INPUT_SIZE:
+        msg = "Unable to set pledged uncompressed content size: %s";
+        break;
+
+    case ERR_LOAD_D_DICT:
+        msg = "Unable to load zstd dictionary or prefix for decompression: %s";
+        break;
+    case ERR_LOAD_C_DICT:
+        msg = "Unable to load zstd dictionary or prefix for compression: %s";
+        break;
+
+    case ERR_GET_C_BOUNDS:
+        msg = "Unable to get zstd compression parameter bounds: %s";
+        break;
+    case ERR_GET_D_BOUNDS:
+        msg = "Unable to get zstd decompression parameter bounds: %s";
+        break;
+    case ERR_SET_C_LEVEL:
+        msg = "Unable to set zstd compression level: %s";
+        break;
+
+    case ERR_TRAIN_DICT:
+        msg = "Unable to train zstd dictionary: %s";
+        break;
+    case ERR_FINALIZE_DICT:
+        msg = "Unable to finalize zstd dictionary: %s";
+        break;
+
+    default:
+        Py_UNREACHABLE();
+    }
+    PyOS_snprintf(buf, sizeof(buf), msg, ZSTD_getErrorName(zstd_ret));
+    PyErr_SetString(state->ZstdError, buf);
+}
+
+typedef struct {
+    const int parameter;
```

Comment:
General comment: we don't use `const T` when `T` is a primitive type such as int even to convene the intent of a constant variable.

Reply:
Is this true in argument positions too? e.g. should I remove the `const Py_ssize_t` from buffer.h?

Reply:
I think we have a bit of both.

* Benefits:
  - avoid mistakes
  - convene the intent of a read-only value
* Downsides:
  - adds more words to read
  - for a function parameter, it does nothing except for a pointer where it makes sense

It's really up to you though, but I would say that it'd be good to be consistent in the `const` usage. I however haven't looked in all places to see if structures are declared with a `const` or not.

But we can discuss this in a separate PR.

Reply:
Thanks for the context. I will think about it. If the tradition is to not have it in functions or structs then I think it would be good to remove them throughout.

Reply:
I removed consts on simple types.